commodore/cluster.py

```diff
@@ -95,7 +103,7 @@ def facts(self) -> dict[str, str]:
 
     @property
     def dynamic_facts(self) -> dict[str, Any]:
-        return self._cluster.get("dynamicFacts", {})
+        return self._cluster.get("dynamicFacts", self._fallback_dynamic_facts)
```

Comment:
If I read this correctly we use the fallback dynamic facts only if the API returns no `dynamicFacts` field, I would probably suggest to also use the fallback facts if the returned fact dict is empty.

The current works with the current API implementation, but as far as I understand the openAPI definition https://github.com/projectsyn/lieutenant-api/blob/master/openapi.yaml returning an empty fact dict would also be valid and sounds like something that could be changed without realizing that it would be a breaking change.

This will most likely never happen, so feel free to ignore this suggestions.

Reply:
I originally explicitly only used the manually provided dynamic facts when there's no field `dynamicFacts` in the API response, to minimize the potential for users to be able to push non-reproducible commits to the cluster catalog by providing incorrect/invalid dynamic facts on the command line for clusters which are bootstrapped but don't report any facts (note that this scenario is currently purely hypothetical as well).

Thinking about it again, I don't really have a strong opinion on whether to allow users to provide arbitrary facts even if the API response for the cluster contains an empty `dynamicFacts` dict.

Reply:
That's also a valid point. I'm not sure anymore which approach is better.

Maybe we could print a warning if Lieutenant returns dynamic facts (event empty ones) and the user tries to provide their own? That way we wouldn't silently fail if the API starts to return `{}`, making debugging a bit easier.